### PR TITLE
More documentation for Semaphore

### PR DESCRIPTION
**Pull request by fjetter**
This extends the documentation a bit further

cc @lr4d

### CODE REVIEW

File: distributed/semaphore.py [hunk 1]
```diff
@@ -204,6 +204,24 @@ def close(self, comm=None, name=None):
 class Semaphore:
     """ Semaphore
 
+    This semaphore will track leases on the scheduler which can be acquired and
```
Comment (lr4d) [inline distributed/semaphore.py:207]:
```suggestion
    This `semaphore <https://en.wikipedia.org/wiki/Semaphore_(programming)>`_
    will track leases on the scheduler which can be acquired and
```

File: distributed/semaphore.py [hunk 2]
```diff
@@ -204,6 +204,24 @@ def close(self, comm=None, name=None):
...
+    released by an instance of this class. If the maximum amount of leases are
+    already acquired, it is not possible to acquire more and the caller waits
+    until another lease has been released.
+
+    The lifetime of a lease are coupled to the ``Client`` it was acquired with.
```
Comment (gforsyth) [inline distributed/semaphore.py:212]:
```suggestion
    The lifetime of a lease is coupled to the ``Client`` it was acquired with.
```

File: distributed/semaphore.py [hunk 3]
```diff
@@ -204,6 +204,24 @@ def close(self, comm=None, name=None):
...
+    Once the Client goes out of scope, the leases associated to it are freed.
+    This behaviour can be controlled with the parameter
```
Comment (gforsyth) [inline distributed/semaphore.py:214]:
```suggestion
    This behavior can be controlled with the parameter
```
both are valid spellings but the rest of the dask docs use the American version

File: distributed/semaphore.py [hunk 4]
```diff
@@ -204,6 +204,24 @@ def close(self, comm=None, name=None):
...
+    `distributed.scheduler.locks.lease-validation-interval`.
+
+    A noticeable difference to the Semaphore of the python standard library is
+    that this implementation does not allow to release more often than it was
+    acquired. If this happens, a warning is emitted but the internal state is
+    not modified.
+
+    This implementation is still in an experimental state and subtle changes in
+    behaviour may occur without any change in the major version of this library.
```
Comment (gforsyth) [inline distributed/semaphore.py:223]:
```suggestion
    behavior may occur without any change in the major version of this library.
```

File: distributed/semaphore.py [hunk 5]
```diff
@@ -204,6 +204,25 @@ def close(self, comm=None, name=None):
...
+    This `semaphore <https://en.wikipedia.org/wiki/Semaphore_(programming)>`_
+    will track leases on the scheduler which can be acquired and
+    released by an instance of this class. If the maximum amount of leases are
+    already acquired, it is not possible to acquire more and the caller waits
+    until another lease has been released.
+
+    The lifetime of a lease is coupled to the ``Client`` it was acquired with.
+    Once the Client goes out of scope, the leases associated to it are freed.
+    This behavior can be controlled with the parameter
+    `distributed.scheduler.locks.lease-validation-interval`.
```
Comment (jrbourbeau) [inline distributed/semaphore.py:216]: Could you update "with the parameter `distributed.scheduler.locks.lease-validation-interval`" to "with the ``distributed.scheduler.locks.lease-validation-interval`` configuration option". Note that we should use two backticks (``) for code highlighting in sphinx

**Comment by dhirschfeld** [comment]
Hi, just read the docs here and wanted to confirm that the `Semaphore` class as-is won't enable [`fire_and_forget`](https://github.com/dask/distributed/blob/66fe0acf041add720d5cc9d9d1ed99d9e90e7e2e/distributed/client.py#L4489) scenarios?

i.e. I would like to spin up a process, fire off 100 tasks to the scheduler using a `Semaphore` to limit the maximum parallelism to 10. The process would then call `fire_and_forget` on the futures and exit.

Once the process which created the tasks exits I'd like them to continue running until completion and to still be bound by the semaphore. From the documentation here it sounds like that once the process exits the tasks will no longer be bound by the semaphore and so will DDOS my database.

Is that correct? i.e. is my desired use-case not supported by the current implementation?

**Comment by dhirschfeld** [comment]
> is my desired use-case not supported by the current implementation

That's fine if so - just trying to understand if I can move my custom code to use a `Semaphore`

**Comment by lr4d** [comment]
@dhirschfeld It should still work fine for your use case. 

> From the documentation here it sounds like that once the process exits the tasks will no longer be bound by the semaphore and so will DDOS my database.

The tasks will still be bound by the semaphore, as the state of the semaphore lies on the scheduler side.
The documentation refers to a semaphore _lease_ being released when the process (referred to as `client`) who acquired the lease disappears. This _lease_ can be thought of as being part of an internal counter at the scheduler side, which ensures that there is an upper bound to the amount of parallelism (i.e. the parameter `max_leases`) you want to allow.

If you want to be sure though, I encourage you to try out the semaphore using `fire_and_forget` for some trivial tasks (e.g. sleeping for a while) and exiting the client before execution has finised, and see what happens

**Comment by lr4d** [comment]
@dhirschfeld Since this feature is still experimental and there may be some (mostly internal) changes upcoming, it would be great if you could contribute a unit test for your use case so that we make sure that it remains supported

**Comment by dhirschfeld** [comment]
Thanks @lr4d - good to know! I just misunderstood the documentation. I'll be giving it a go as soon as I can find the time and am happy to report back if everything works as expected... or otherwise :)

**Pull request merged**